lighthouse-core/lib/page-functions.js

```diff
@@ -163,14 +163,25 @@ function ultradumbBenchmark() {
 
 /**
  * Adapted from DevTools' SDK.DOMNode.prototype.path
- *   https://github.com/ChromeDevTools/devtools-frontend/blob/7a2e162ddefd/front_end/sdk/DOMModel.js#L530-L552
- * TODO: Doesn't handle frames or shadow roots...
+ *   https://github.com/ChromeDevTools/devtools-frontend/blob/4fff931bb/front_end/sdk/DOMModel.js#L625-L647
+ * Backend: https://source.chromium.org/search?q=f:node.cc%20symbol:PrintNodePathTo&sq=&ss=chromium%2Fchromium%2Fsrc
+ *
+ * TODO: DevTools nodePath handling doesn't currently support iframes, but probably could.
  * @param {Node} node
  */
 /* istanbul ignore next */
 function getNodePath(node) {
+  // For our purposes, there's no worthwhile difference between shadow root and document fragment
+  // We can consider them entirely synonymous.
+  const isShadowRoot = node => node.nodeType === Node.DOCUMENT_FRAGMENT_NODE;
+  const getNodeParent = node => isShadowRoot(node) ? node.host : node.parentNode;
+
   /** @param {Node} node */
   function getNodeIndex(node) {
+    if (isShadowRoot(node)) {
+      // User-agent shadow roots get 'u'. Non-UA shadow roots get 'a'.
+      return 'a';
```

Comment:
This is the most key thing. the `index` value for shadow roots has to be `a`. Why? i have no idea. But it won't work with anything else.

An added challenge here is that in order to fully test nodePath we gotta verify that the devtools backend and frontend can successfully consume a nodePath that Lighthouse created.  We currently dont have any automated way to do that. 

I did it manually to verify that the string our method here generates matches exactly what devtools does.. (and that it resolves in the devtools backend).

Reply:
> Why? I have no idea.

Is this because its a behaviour defined in Chrome? as in we're just complying with whatever the implementation is over there basically

Reply:
heh yes thats right. this nodePath thing was just invented by some chrome devtools folks years ago. the choice of 'u'/'a' was also seemingly arbitrary, so yeah.. we're just matching to make sure we remain compatible. 

there's some comment somewhere saying they should use xpath instead of this invented nodePath thing by no one ever got around to it.